brontide/read_pool.go

```diff
@@ -0,0 +1,179 @@
+// channel for returning errors to the caller.
+type readRequest struct {
+	fn      func(*ReadBuffer) error
+	errChan chan error
+}
+
+// ReadPool is a worker pool that can be used to read, decrypt, and deserialize
+// lnwire messages and from the wire. The ReadPool schedules the execution of
+// function closures accepting a *ReadBuffer. The ReadPool uses a backing
+// ReadBufferPool to conserve allocations related to creating ReadBuffers, and
+// also serves to decouple these allocations from the connection life cycle.
+type ReadPool struct {
+	stopped uint32 // to be used atomically
+
+	timeout time.Duration
+	pool    *ReadBufferPool
+
+	work      chan *readRequest
+	workerSem chan struct{}
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+// NewReadPool initializes a new ReadPool using bufferPool as the backing pool
+// of ReadBuffers. At most numWorkers goroutines will be spawned to process
+// requests, each of which is independently released if it receives no tasks for
+// a consecutive duration of workerTimeout.
+func NewReadPool(bufferPool *ReadBufferPool, numWorkers int,
+	workerTimeout time.Duration) *ReadPool {
+	return &ReadPool{
+		timeout:   workerTimeout,
+		pool:      bufferPool,
+		work:      make(chan *readRequest),
+		workerSem: make(chan struct{}, numWorkers),
+		quit:      make(chan struct{}),
+	}
+}
+
+// Stop safely shuts down the ReadPool.
+func (p *ReadPool) Stop() {
+	if !atomic.CompareAndSwapUint32(&p.stopped, 0, 1) {
+		return
+	}
+
+	close(p.quit)
+	p.wg.Wait()
+}
+
+// Submit takes a function closure and submits it to a worker for processing.
+// The function will be invoked with a clean *ReadBuffer with enough capacity
+// for any _encrypted_ lnwire message. If successful, this method blocks until
+// the function has been invoked. Otherwise, ErrReadPoolExiting is returned
+// signaling the task could not be submitted because the ReadPool is shutting
+// down.
+func (p *ReadPool) Submit(fn func(*ReadBuffer) error) error {
+	req := &readRequest{
+		fn:      fn,
+		errChan: make(chan error),
+	}
+
+	select {
+
+	// If we have not reached our maximum number of workers, spawn one to
+	// process the submitted request.
+	case p.workerSem <- struct{}{}:
+		p.wg.Add(1)
+		go p.spawnWorker(req)
+
+	// Otherwise, submit the task to any of the active workers.
+	case p.work <- req:
+
+	case <-p.quit:
+		return ErrReadPoolExiting
+	}
+
+	// Wait for task to be processed. We don't select on quit here because
+	// we want to ensure we only return after the function has finished.
+	return <-req.errChan
+}
+
+// spawnWorker creates a worker goroutine that processes the initial task it is
+// given, then attempts to take new requests from the pool's work channel. If no
+// requests are received before the worker's timeout expires, the worker will
+// exit and return the buffer it was using to the buffer pool.
+//
+// NOTE: This method MUST be run as a goroutine.
+func (p *ReadPool) spawnWorker(req *readRequest) {
```

Comment:
The code overlap between this and the write pool is pretty substantial. I can think of a few ways to consolidate the two, but most of them involve either the addition of an `interface{}` here and there or additional type assertions. Both these methods would incur an additional performance hit as they need to be converted back into concrete types at the end points...